iquip/apps/logger.py

```diff
@@ -0,0 +1,189 @@
+"""Module for log viewer in apps."""
+
+import time
+import logging
+from typing import Any, Optional, Tuple, Callable
+
+from PyQt5.QtCore import QObject, pyqtSlot, pyqtSignal
+from PyQt5.QtWidgets import (
+    QWidget, QVBoxLayout, QPushButton, QTextEdit, QLabel, QDialogButtonBox, QComboBox
+)
+
+import qiwis
+
+logger = logging.getLogger(__name__)
+
+
+class _Signaller(QObject):
+    """Signal only for LoggingHandler.
+
+    Signals:
+        signal(log): A formatted log message is emitted. 
+    """
+
+    signal = pyqtSignal(str)
+
+
+class LoggingHandler(logging.Handler):
+    """Handler for logger.
+
+    Sends a log message to connected function using emit.   
+    """
+
+    def __init__(self, slotfunc: Callable[[str], Any]):
+        """Extended.
+
+        Connects the slotfunc to the signal.
+
+        Args:
+            slotfunc: A slot function which is called when a log record is emitted.
+        """
+        super().__init__()
+        self.signaller = _Signaller()
+        self.signaller.signal.connect(slotfunc)
+
+    def emit(self, record: logging.LogRecord):
+        """Overridden.
+        
+        Emits input signal to the connected function.
+        """
+        s = self.format(record)
+        self.signaller.signal.emit(s)
+
+
+class LoggerFrame(QWidget):
+    """Frame for logging.
+
+    Attributes:
+        logEdit: A textEdit which shows all logs.
+        clearButton: A button for clearing all logs.
+        levelBox: A comboBox for setting the logger's level.
+    """
+
+    def __init__(self, parent: Optional[QObject] = None):
+        """Extended."""
+        super().__init__(parent=parent)
+        # widgets
+        self.logEdit = QTextEdit(self)
+        self.logEdit.setReadOnly(True)
+        self.clearButton = QPushButton("Clear")
+        self.levelBox = QComboBox(self)
+        # layout
+        layout = QVBoxLayout(self)
+        layout.addWidget(self.logEdit)
+        layout.addWidget(self.clearButton)
+        layout.addWidget(self.levelBox)
+
+
+class ConfirmClearingFrame(QWidget):
+    """
+    A confirmation frame for log clearing.
+    """
+    confirmed = pyqtSignal()
+
+    def __init__(self, parent: Optional[QObject] = None):
+        """
+        Extended.
+        """
+        super().__init__(parent=parent)
+        # widgets
+        self.label = QLabel("Are you sure to clear?")
+        self.buttonBox = QDialogButtonBox()
+        self.buttonBox.addButton("OK", QDialogButtonBox.AcceptRole)
+        self.buttonBox.addButton("Cancel", QDialogButtonBox.RejectRole)
+        # connect signals
+        self.buttonBox.accepted.connect(self.buttonOKClicked)
+        self.buttonBox.rejected.connect(self.buttonCancelClicked)
+        # layouts
+        layout = QVBoxLayout()
+        self.setLayout(layout)
+        layout.addWidget(self.label)
+        layout.addWidget(self.buttonBox)
+
+    def buttonOKClicked(self):
+        """Clicks OK to clear log."""
```

Comment:
It seems like this method is clicking the OK button.
How about "Called when the OK button is clicked."?

Reply:
It seems reasonable. I will also change same part in qiwis.